Add MLE/MAP functionality

cpfiffer opened the pull request:
This PR adds MLE and MAP functions to generate point estimates. I have not yet written tests, so this is still a WIP.

Usage:

```julia
@model function gdemo(x)
    s ~ InverseGamma(15,3)
    m ~ Normal(0, 1)

    for i in eachindex(x)
        x[i] ~ Normal(m, s)
    end

    return m, s
end

Random.seed!(100)
data = rand(Normal(0, 2), 10)
model = gdemo(data)

# Make point estimates
mle_gdemo = MLE(model)
map_gdemo = MAP(model)
```

MLE results:

```julia
2×3 Named Array{Float64,2}
A ╲ B │ parameter    std_err      tstat
──────┼────────────────────────────────
s     │   2.08011   0.465139    4.47202
m     │  0.596312   0.657789    0.90654
```

MAP results:

```julia
2×3 Named Array{Float64,2}
A ╲ B │ parameter    std_err      tstat
──────┼────────────────────────────────
s     │    1.3783    0.20177    6.83104
m     │  0.501096    0.40024    1.25199
```

mohamed82008 commented:
I suppose that for many data points, `m` converges to 0 in MLE?

cpfiffer commented:
Yeah, the MLE estimate here is basically the mean.

devmotion commented:
Honestly, I had no clue what you mean with standard errors and tstat before I checked the code. I would have expected it to just return the mode, either as named tuple or VarInfo object, possibly together with the other results from the optimizer. Maybe one could have a separate function for these diagnostics?

If `mode_estimation` "just" performs the optimization, an alternative approach could also be to just provide the cost function and initial value to the user, similar to https://github.com/baggepinnen/FluxOptTools.jl, so users can just call Optim (or possibly other optimization packages) in whatever way without Turing having to worry about it.

devmotion commented on Project.toml:
```diff
@@ -17,6 +17,7 @@ Libtask = "6f1fad26-d15e-5dc8-ae53-837a1d7b8c9f"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 LogDensityProblems = "6fdf6af0-433a-55f7-b3ed-c6c6e0b8df7c"
 MCMCChains = "c7f686f2-ff18-58e9-bc7b-31028e88f75d"
+Optim = "429524aa-4258-5aef-a3af-852621145aeb"
```
Maybe make this an optional requirement?

devmotion replied:
These dependencies should be moved to the `[extras]` section, since the MAP/MLE functionality is optional.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+module ModeEstimation
```
I guess this is motivated by the fact that it might become a separate package at some point? IMO having nested modules is usually not very convenient for handling imports and exports (and debugging...), so I try to avoid it.

cpfiffer replied:
Well, we only export two things, and I'm hoping this is not going to live in Turing for very long, so I wanted to keep all of these things separate.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+
+A wrapper struct to store various results from a MAP or MLE estimation.
+
+Fields:
+
+- `values` is a vector with the resulting point estimates
+- `info_matrix` is the inverse Hessian
+- `optim_result` is the stored Optim.jl results
+- `summary_table` is a summary table with parameters, standard errors, and 
+  t-statistics computed from the information matrix.
+- `lp` is the final likelihood.
```
I guess it would be nicer (and more standard) to add the documentation directly to the fields and use DocStringExtensions to include them in the docstring (see https://github.com/TuringLang/Turing.jl/pull/1131).

cpfiffer replied:
Nice, I hadn't noticed that. I'll add it.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+"""
+    make_logjoint(model::DynamicPPL.Model, ctx::DynamicPPL.AbstractContext)
+
+Constructs a log density function that accepts a vector `z` and returns 
+a tuple (-likelihood, `varinfo`). The model is run using the provided 
+context `ctx`.
```
The [first one-line sentence in each docstring should use the imperative form](https://docs.julialang.org/en/latest/manual/documentation/):
```suggestion
Construct a log density function.

The `model` is run using the provided context `ctx`.
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+a tuple (-likelihood, `varinfo`). The model is run using the provided 
+context `ctx`.
+"""
+function make_logjoint(model::DynamicPPL.Model, ctx::DynamicPPL.AbstractContext)
+    # setup
+    varinfo_init = Turing.VarInfo(model)
+    spl = DynamicPPL.SampleFromPrior()    
+    DynamicPPL.link!(varinfo_init, spl)
+
+    function logπ(z; unlinked = false)
+        varinfo = DynamicPPL.VarInfo(varinfo_init, spl, z)
+
+        unlinked && DynamicPPL.invlink!(varinfo_init, spl)
+        model(varinfo, spl, ctx)
+        unlinked && DynamicPPL.link!(varinfo_init, spl)
+
+        return -DynamicPPL.getlogp(varinfo)
+    end
+
+    return logπ
+end
```
I know this pattern is used all over the place in Turing but IMO it would be cleaner to just have a proper callable struct, something like
```julia
struct OptimLogDensity{M<:Model,C<:Context,V<:VarInfo}
	model::M
	context::C
	init::V
end

function OptimLogDensity(model::Model, context::AbstractContext)
	init = VarInfo(model)
	DynamicPPL.link!(init, SampleFromPrior())
	return OptimLogDensity(model, context, init)
end

function (f::OptimLogDensity)(z)
	...
end
```

That would make it easier to access `model`, `context`, and the initial sample (which is created multiple times currently it seems) from the log density function.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+function mode_estimation(
+    model::DynamicPPL.Model, 
+    lpf; 
+    optim_options=Optim.Options(),
```
Maybe just forward `kwargs...` to Optim instead? Or just provide the cost function that can be optimized alternatively.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+    init_params = model(vi, spl)
+    init_vals = vi[spl]
+
+    # Construct target function.
+    target(x) = lpf(x)
+    hess_target(x) = lpf(x; unlinked=true)
+
```
There's no need to construct these functions. Instead of `target` one can just use `lpf`, and the second one can be avoided by using `do` syntax when computing the Hessian (which maybe should not be part of this function?).

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+
+"""
+function mode_estimation(
+    model::DynamicPPL.Model, 
+    lpf; 
```
There is no need to provide `model` if `lpf` would be a callable struct, as mentioned above.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+    binv = inv(b)
+
+    spl = DynamicPPL.SampleFromPrior()
+    vi = DynamicPPL.VarInfo(model)
+    init_params = model(vi, spl)
```
I guess these model evaluations could be avoided by reusing the sample that was computed when constructing the log density function.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+    model(vi) # XXX: Is this a necessary step?
+
+    # Make one transition to get the parameter names.
+    ts = [Turing.Inference.Transition(DynamicPPL.tonamedtuple(vi), DynamicPPL.getlogp(vi))]
+    varnames, _ = Turing.Inference._params_to_array(ts)
```
Can't you get the parameter names from `VarInfo` directly? If not, we should change that since it seems weird and inefficient to first convert it to a Transition and then extract the names.

cpfiffer replied:
Not in a quick way, at least as far as I can tell. If someone's got a good fix other than this (reasonbly good) hack, I'm game.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+    # Retrieve the estimated values.
+    vals = binv(M.minimizer)
+
+    # Get the VarInfo at the MLE/MAP point, and run the model to ensure 
+    # correct dimensionality.
+    vi[spl] = vals
+    model(vi) # XXX: Is this a necessary step?
```
Do you have an example for what could go wrong when the model is not rerun?

cpfiffer replied:
Not really. I pulled this out.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+        kwargs...
+    )
+
+An internal function that handles the computation of a MLE or MAP estimate.
```
Being picky about the imperative form again, maybe something like:
```suggestion
Estimate a mode, i.e., compute a MLE or MAP estimate.
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+
+An internal function that handles the computation of a MLE or MAP estimate.
+
+Arguments: 
+
+- `model` is a `DynamicPPL.Model`.
+- `lpf` is a function returned by `make_logjoint`.
+
+Optional arguments:
+
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
+
```
Argument lists [should be avoided if possible](https://docs.julialang.org/en/latest/manual/documentation/).

devmotion commented on src/modes/ModeEstimation.jl:
````diff
@@ -0,0 +1,235 @@
+mle_estimate = MLE(model, optim_options=Optim.Options(iterations=500))
+```
+"""
+function MLE(model::DynamicPPL.Model; kwargs...)
````
IMO this should be
```suggestion
function mle(model::DynamicPPL.Model; kwargs...)
```
since `MLE` suggests incorrectly that an object of type `MLE` is returned.

devmotion replied:
Alternatively, one could use singleton types `MLE` and `MAP` as inputs to `mode_estimation` instead of the cost function.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+"""
+    MLE(model::DynamicPPL.Model; kwargs...)
+
+Returns a maximum likelihood estimate of the given `model`.
```
```suggestion
Compute a maximum likelihood estimate of the given `model`.
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+
+Returns a maximum likelihood estimate of the given `model`.
+
+Arguments: 
+
+- `model` is a `DynamicPPL.Model`.
+
+Keyword arguments:
+
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
```
Again, maybe avoid these lists and just forward regular keyword arguments to Optim?

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
+
+Usage:
```
Code examples should be included in an [`# Examples` section](https://docs.julialang.org/en/latest/manual/documentation/):
```suggestion
# Examples
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+"""
+    MAP(model::DynamicPPL.Model; kwargs...)
+
+Returns the maximum a posteriori estimate of the given `model`.
```
```suggestion
Compute the maximum a posteriori estimate of the given `model`.
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
+
+Usage:
```
```suggestion
# Examples
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,235 @@
+
+Returns the maximum a posteriori estimate of the given `model`.
+
+Arguments: 
+
+- `model` is a `DynamicPPL.Model`.
+
+Keyword arguments:
+
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
```
Same as above, maybe try to avoid the explicit lists.

devmotion commented on src/modes/ModeEstimation.jl:
````diff
@@ -0,0 +1,235 @@
+mle_estimate = MAP(model, optim_options=Optim.Options(iterations=500))
+```
+"""
+function MAP(model::DynamicPPL.Model; kwargs...)
````
This should be a function as well but unfortunately it can't be named `map`... Maybe use something more explicit?

cpfiffer commented:
> Honestly, I had no clue what you mean with standard errors and tstat before I checked the code. I would have expected it to just return the mode, either as named tuple or VarInfo object, possibly together with the other results from the optimizer. Maybe one could have a separate function for these diagnostics?

No, this is almost completely counter to how these analyses are supposed to work. Using a MLE/MAP estimate with no information on the dispersion of that estimator is next to useless, akin to running OLS without standard errors or using just the median or whatever from a full posterior.

I think if you are the kind of person who works in the MLE world, it's an absolute requirement to have statistics computed from the information matrix. Plus, you can use it to run hypothesis tests.

Granted, perhaps this should be separated out, such that you do something like

```julia
est = MLE(model)
info_matrix = information_matrix(est)
```

Or something equivalent. 

But I would prefer to hew to how OLS works, where the estimator covariance matrix is returned simultaneously with the point estimates.

devmotion commented:
> Granted, perhaps this should be separated out, such that you do something like
>
> est = MLE(model)
> info_matrix = information_matrix(est)
>
> Or something equivalent.

Yes, that's what I had in mind (I just thought you might have to provide the `model` to `information_matrix` as well). IMO separating these computations seems cleaner, even if you are always interested in the information matrix, since it would allow to evaluate the information matrix also for estimates that are obtained in a different way. And, of course, one can still implement a function that just calls both functions and returns their results, if one wants to save one line in downstream code.

yebai commented:
A minor comment on the interface: maybe we can create a new `optimise(model, algorithm)` API on top of the `sample(model, algorithm)` method. This `optimise` function can support both point estimation and variational inference algorithms, e.g. 

```julia
# Make point estimates
mle_gdemo = optimise(model, MLE(...))
map_gdemo = optimise(model, MAP(...))
vi_gdemo = optimise(model, ADVI(...))
```

cc @torfjelde

cpfiffer commented:
One quibble with this is that `optimize` would return two different things -- the MLE/MAP return values are distinct from the ADVI stuff in that it is a vector of point estimates, which I think is a little counter to the interface method where we generally have a framework for the return value.

yebai commented:
> One quibble with this is that optimize would return two different things -- the MLE/MAP return values are distinct from the ADVI stuff in that it is a vector of point estimates, which I think is a little counter to the interface method where we generally have a framework for the return value.

That's true, but we can probably introduce some abstract types which can be shared by all deterministic inference algorithms -- point estimation returns a Dirac measure, which can be viewed as a special variational distribution.

cpfiffer commented:
Okay, I've changed it quite a bit. There's now a callable struct `OptimLogDensity` which gets you the log joint or the log likelihood, as per David's suggestion.

I have also implemented the `Optim.optimize` call @yebai suggested, such that you now do

```julia
using Turing
using Optim

@model function gdemo(x)
    s ~ InverseGamma(15,3)
    m ~ Normal(0, 1)

    for i in eachindex(x)
        x[i] ~ Normal(m, s)
    end

    return m, s
end

Random.seed!(100)
data = rand(Normal(0, 2), 10)
model = gdemo(data)

mle = optimize(model, MLE())
mpa = optimize(model, MAP())
```

I also implemented a bunch of StatsBase.jl functions to handle the information matrix, covariance matrix, etc. It's much more palatable now, I think. 

Currently the MAP estimates are incorrect, and I can't figure out why -- they are biased downards by quite a bit. The current numbers for the default `gdemo` model are

```julia
A  │                                                                                                                           
───┼────────                                                                                                                   
s  │ 1.16668                                                                                                                   
m  │ 1.16668    
```

where we expect the results to be

```julia
A  │                                                                                                                           
───┼────────                                                                                                                   
s  │ 2.0416                                                                                                                   
m  │ 1.16666  
```

The fact that the mean is correct while the variance is not is making me suspect there's an issue with the bijection somewhere, which I'll have to look into later.

trappmartin commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,205 @@
+
+Estimate a mode, i.e., compute a MLE or MAP estimate.
+"""
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.BFGS(), args...; kwargs...)
```
The memory requirement of BFGS scales quadratic in the number of parameter you optimise. I would suggest to use a more memory friendly option as default.
```suggestion
function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim. LBFGS(), args...; kwargs...)
```
or 
```suggestion
function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.GradientDescent(), args...; kwargs...)
```

devmotion commented:
Could be interesting to use [GalacticOptim.jl](https://github.com/sciml/galacticoptim.jl) at some point.

trappmartin commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,205 @@
+
+    varinfo = DynamicPPL.VarInfo(f.init, spl, z)
+
+    unlinked && DynamicPPL.invlink!(f.init, spl)
```
Do I understand it correctly that this is done in each iteration. Is it really necessary to check this at each iteration, or could you pull out the code instead and do this once?

trappmartin commented:
This would be really cool! GalacticOptim.jl seems like the goto in the future.

devmotion commented:
Just throwing ideas around: a different approach to support optimization would be something like https://github.com/baggepinnen/FluxOptTools.jl, i.e., just providing a function that provides users with the functions that they plug into Optim (or other optimization packages).

cpfiffer commented:
If you want just the density to use somewhere, you can do that with

```julia
@model function f(x)
    m ~ Normal()
    x ~ Normal(m,1)
end

model = f(1.5)
target = Turing.OptimLogDensity(model, DynamicPPL.LikelihoodContext())
```

but perhaps there's a convenience function we could include as a first-class citizen in Turing -- maybe `logdensity(model)` and `loglikelihood(model)`.

trappmartin commented:
> but perhaps there's a convenience function we could include as a first-class citizen in Turing -- maybe `logdensity(model)` and `loglikelihood(model)`.

I like the idea. But maybe better: `logjoint(model)` and `loglikelihood(model)`, log density is not very descriptive as it doesn't say what the density is of.

cpfiffer commented:
Alright. I need some other eyes on this one. Something is wrong with the transformations, but I cannot figure out where -- MAP yields a result of

```julia
ModeResult with minimized lp of -4.60

2-element Named Array{Float64,1}
A  │ 
───┼────────
s  │  1.1667
m  │ 1.16667
```

when it should be 

```julia
2-element Named Array{Float64,1}
A  │ 
───┼────────
s  │  2.0416 
m  │ 1.16667
```

1.1667 is quite a bit lower than 2.0416, and I know it's because the transforms are wrong. I've tried checking to see if there's a missing inverse bijection (I don't think there is), if I'm using `link!` and `invlink!` correctly (not sure that I am, but couldn't figure out a different approach to fix it). 

Box minimization works fine, but I'd rather not add the computational overhead of figuring out bounds dynamically when transforms exist. 

Here's a MWE:

```julia
using Random
using Turing
using Optim

@model function gdemo()
    s ~ InverseGamma(2,3)
    m ~ Normal(0, sqrt(s))

    1.5 ~ Normal(m, sqrt(s))
    2.0 ~ Normal(m, sqrt(s))

    return m, s
end

Random.seed!(100)
model = gdemo()

mle = optimize(model, MLE())
mpa = optimize(model, MAP())

mle |> display
mpa |> display
```

trappmartin commented:
I'll see if I can find the problem.

trappmartin commented:
@cpfiffer should the code you posted run on this PR? Seem like `NamedArrays` is missing.
```
ERROR: LoadError: LoadError: ArgumentError: Package Turing does not have NamedArrays in its dependencies:
- If you have Turing checked out for development and have
  added NamedArrays as a dependency but haven't updated your primary
  environment's manifest file, try `Pkg.resolve()`.
- Otherwise you may need to report an issue with Turing
```

Could you check if everything is committed.

cpfiffer commented:
Should be good to go with the recent fixes.

trappmartin commented:
Thx

cpfiffer commented:
Okay, I took another crack at trying to figure this out. I do _not_ have an answer, but I'm wondering if the evidence I've got might help a little. Here's my MWE (using this branch):

```julia
using Bijectors
using Turing
using Optim

@model function gdemo()
    s ~ InverseGamma(2,3)
    m ~ Normal(0, sqrt(s))

    1.5 ~ Normal(m, sqrt(s))
    2.0 ~ Normal(m, sqrt(s))

    return m, s
end

model = gdemo()

bm = bijector(model)

globmax = [49/24, 7/6]

@info "Unlinked"
mpa_fun(globmax, linked=false) |> display

@info "Linked"
mpa_fun(bm(globmax), linked=true) |> display
```

The issue is basically that, even if I biject the true MAP estimate of `[49/24, 7/6]` and then have Turing link it internally, I get a result I'm not sure I understand. This is perhaps generally in line with my inability to grok bijectors. Here's the output:

```julia
[ Info: Unlinked

[2.0416666666666665, 1.1666666666666667]
(s = ([2.0416666666666665], ["s"]), m = ([1.1666666666666667], ["m"]))
Assume s: val: 2.0416666666666665 lp: -1.4134625810538644
Assume m: val: 1.1666666666666667 lp: -1.6091551004193465
Unlinked LP: 5.771540127209862
```

```julia
[ Info: Linked

[0.7137664677626809, 1.1666666666666667]
(s = ([2.0416666666666665], ["s"]), m = ([1.1666666666666667], ["m"]))
Assume s: val: 2.0416666666666665 lp: -0.6996961132911834
Assume m: val: 1.1666666666666667 lp: -1.6091551004193465
Linked LP: 5.057773659447182
```

My question: Why do I get different densities if I biject the true result first and let Turing handle the link internally, vs. using the unlinked variables with no transform? Shouldn't they be equivalent?

Maybe @torfjelde might have an idea -- I need an adult.

trappmartin commented:
The logpdf of the transformed density function is only for trivial cases the same as the untransformed density function, i.e. affine transformations. What is important is that both share the same maximum, i.e. the transformation has to be monotonic (I think). 

On a different note. 
I had a look at the PR a few days back and checked it works for the prior context, which it did. I haven't had the time to continue but it might be good to change the PR so that it uses the latest merged PR from DynamicPPL, which allow us to use `loglikelihood(model, vi)`. Thus, quite a bit of the code in this PR could be removed I believe. What do you think?

cpfiffer commented:
> The logpdf of the transformed density function is only for trivial cases the same as the untransformed density function, i.e. affine transformations. What is important is that both share the same maximum, i.e. the transformation has to be monotonic (I think).

They have different maxima. The untransformed MAP estimate is (correctly) 2.0416, whereas the transformed MAP estimate is 1.6667. So something is wrong with the transform?

> On a different note.
> I had a look at the PR a few days back and checked it works for the prior context, which it did. I haven't had the time to continue but it might be good to change the PR so that it uses the latest merged PR from DynamicPPL, which allow us to use `loglikelihood(model, vi)`. Thus, quite a bit of the code in this PR could be removed I believe. What do you think?

I don't think it would actually eliminate much code at all, since much of the callable struct functionality handles linking and contexts that those convenience functions don't.

trappmartin commented:
I just created a tiny gist, as I find it more convenient to implement this differently and try out stuff there. Actually, even the maximisation for the prior does not converge to the mode. No idea what the problem is atm.

https://gist.github.com/trappmartin/597b3de3d348006415f61c325f006657

cpfiffer commented:
Well, at the very least I'm happy that your method is also finding some strange behavior.

devmotion commented:
As far as I can see, the following is happening here when you do not convert the values back to the untransformed space:
- You update the `VarInfo` object (which has `trans` flag set to `true` for all variables since you called `link!` before) with the transformed values `xtrans` by calling `setval!`
- During model execution, the untransformed value `x` is retrieved when evaluating the assume statements: https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/context_implementations.jl#L134 and https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/varinfo.jl#L800-L801
- Then https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/context_implementations.jl#L141 evaluates `logpdf_with_trans(dist, x, true)` which just returns `logpdf(dist, x) - logabsdetjac(bijector(dist), x)`: https://github.com/TuringLang/Bijectors.jl/blob/1f3b581afe04f690bd93fba9edd88735cc1fc140/src/Bijectors.jl#L147-L153

However, if you convert the values manually to the untransformed space, the following happens:
- You update the `VarInfo` object (which has `trans` flag set to `false` for all variables since you did not call `link!` before) with the untransformed values `x` by calling `setval!`
- During model execution, the untransformed value `x` is retrieved when evaluating the assume statements: https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/context_implementations.jl#L134 and https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/varinfo.jl#L800-L801
- Then https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/context_implementations.jl#L141 evaluates `logpdf_with_trans(dist, x, false)` which just returns `logpdf(dist, x)`: https://github.com/TuringLang/Bijectors.jl/blob/1f3b581afe04f690bd93fba9edd88735cc1fc140/src/Bijectors.jl#L147-L153

So it is not surprising that you not obtain the same log densities in both cases. The optimization procedure should be unaffected by this difference only if the additional `logabsdetjac` term does not affect the location of the maximum (or minimum, depending on what you optimize). I don't know why that's not the case here and if bijectors are supposed to have this property in general at all.

However, since the values are converted to the untransformed space in DynamicPPL anyways during model evaluation, maybe one should just manually call `invlink!` after updating the values with `setval!` to get rid of the `logabsdetjac` term?

trappmartin commented:
I think all of the current steps should be fine. The transformation might effect the optimisation in terms of how well the optimiser finds the maximum but the maximum should not change given the transformation that is applied. I’ll look at it tomorrow again, it’s later over here. That said, @cpfiffer are actually comparing against the mode of the true posterior. I just did some calculations in my head (not reliable) and I think 2 sounds more like the mean of the posterior of the inverse gamma then the mode to me. I think the mode should be smaller. But again, it’s late over here and I’ll check again tomorrow.

devmotion commented:
> The transformation might effect the optimisation in terms of how well the optimiser finds the maximum but the maximum should not change given the transformation that is applied.

I just did the calculations on pen and paper for the prior example that you posted above, where optimization yielded an optimal value for `s` of around 1.2 whereas we would expect a value of 6/7, which is approx 0.86. Here's my calculation (constants denoted by C_1 and C_2):
1. We have `logpdf(IG(2, 3), s) = C_1 - (2 + 1) * log(s) - 3 / s = C_1 - 3 log(s) - 3 / s`
2. We have `logpdf(Normal(0, sqrt(s)), m) = C_2 - (log(s) + m^2 / s) / 2 = C_2 - log(s) / 2 - m^2 / (2 s)`
3. Thus in total we get `logpdf(IG(2, 3), s) + logpdf(Normal(0, sqrt(s)), m) = C_1 + C_2 - 7 log(s) / 2 - (6 + m^2) / (2 s)`
4. This expression is maximized for `m = 0` and `s = 6 / 7`.
5. The bijectors that are used in this case are the identity for `m` and `log` for `s`, and hence we have `logabsdetjac(bijector(IG(2, 3)), s) = - log(s)` (see https://github.com/TuringLang/Bijectors.jl/blob/d626f046b7f49160e22d597d2df6231413428180/src/bijectors/exp_log.jl#L42) and `logabsdetjac(bijector(Normal(0, sqrt(s)), m) = 0` (see https://github.com/TuringLang/Bijectors.jl/blob/a763e3ff922743d0cf2a0c6260e9ec5926d3adc9/src/interface.jl#L116).
6. Thus with the bijectors we obtain `logpdf(IG(2, 3), s) - logabsdetjac(bijector(IG(2, 3)), s) + logpdf(Normal(0, sqrt(s)), m) - logabsdetjac(bijector(Normal(0, sqrt(s))), m) = C_1 + C_2 - 7 log(s) / 2 - (6 + m^2) / (2 s) - (-log(s)) = C_1 + C_2 - 5 log(s) / 2 - (6 + m^2) / (2 s)`.
7. This expression is maximized for `m = 0` and `s = 6 / 5`.

Thus the optimization based on the transformed values (with the additional terms of `-logabsdetjac`) correctly returned a value close to 1.2. However, that maximum is different from the one you would obtain by performing optimization based on the `logpdf` of the untransformed values only.

mohamed82008 commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,210 @@
+
+    varinfo = DynamicPPL.VarInfo(f.init, spl, z)
+
+    if unlinked
```
I think this makes more sense to be a field of `OptimLogDensity`. When finding the information matrix, we can construct a new `OptimLogDensity` with `unlinked = true` or `linked = false`.

mohamed82008 commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,210 @@
+
+Estimate a mode, i.e., compute a MLE or MAP estimate.
+"""
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
```
Could be nice to allow an initial solution here as a named tuple using PriorContext.

cpfiffer replied:
Do you mean as a return value or a starting point?

devmotion replied:
I think that should be part of the `OptimLogDensity` constructor ideally and not be handled here. For convenience, one could then forward it from the `MAP` and `MLE` methods (or include it as a field?) when `OptimLogDensity` is constructed.

mohamed82008 replied:
>Do you mean as a return value or a starting point?

I mean as a starting point. We also need to make sure the varinfo here is linked. If the same objective is optimized twice, the varinfo will not be linked here I think.

cpfiffer replied:
You're right on that -- I've linked it after the optimized values are extracted. Good catch.

devmotion commented:
The calculation for the MAP estimate is very similar, we just get the additional terms `logpdf(Normal(m, sqrt(s)), 1.5)` and `logpdf(Normal(m, sqrt(s)), 2)` (since the bijector of the normal distribution is the identity, we can neglect it). Then without transformations we get the expression
`C_1 + C_2 - 7 log(s) / 2 - (6 + m^2) / (2 s) - log(s) / 2 - (m - 1.5)^2 / (2 s) - log(s) / 2 - (m - 2)^2 / (2 s) = C_1 + C_2 - 9 log(s) / 2 - (6 + m^2 + (m - 1.5)^2 + (m - 2)^2) / (2 s)`, which has the maximum `m = 7 / 6` and `s = 49 / 54`. With the transformations, and hence with the additional `logabsdetjac` term, we get the expression
`C_1 + C_2 - 5 log(s) / 2 - (6 + m^2) / (2 s) - log(s) / 2 - (m - 1.5)^2 / (2 s) - log(s) / 2 - (m - 2)^2 / (2 s) = C_1 + C_2 - 7 log(s) / 2 - (6 + m^2 + (m - 1.5)^2 + (m - 2)^2) / (2 s)`, which is maximized for `m = 7 / 6` and `s = 7 / 6`.

Thus the optimization gave the correct answer for the transformed values. It's just not the same maximum that you would get for the untransformed values.

mohamed82008 commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,210 @@
+    binv = inv(b)
+
+    spl = DynamicPPL.SampleFromPrior()
+    vi = DynamicPPL.VarInfo(model)
```
Can we use the VarInfo inside `f`?

mohamed82008 commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,210 @@
+    init_vals = vi[spl]
+
+    # Optimize!
+    M = Optim.optimize(f, init_vals, args...; kwargs...)
```
`init_vals` here will be in the constrained domain. Using the `VarInfo` inside `f` should fix this since it is linked.

mohamed82008 commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,210 @@
+"""
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+    # Do some initialization.
+    b = bijector(model)
```
I don't think we need this. We can just write the minimizer to `f.init`, `invlink!` it, get the solution in the constrained space, then `link!` it back.

mohamed82008 commented:
I agree with @devmotion's analysis. I think we should have an `OptimizationContext` that always evaluates `logpdf` instead of `logpdf_with_trans` but still calls `invlink` in `vi[vn]`, i.e. `istrans(vi, vn)` is true but `logpdf` is called and not `logpdf_with_trans`.

devmotion commented:
BTW it's new to me but apparently it's a well-known fact that the MAP estimate are not invariant under reparameterization: https://stats.stackexchange.com/questions/330402/what-is-an-example-of-a-transformation-on-a-posterior-distribution-such-that-the

devmotion commented:
> I think we should have an OptimizationContext that always evaluates logpdf instead of logpdf_with_trans but still calls invlink in vi[vn], i.e. istrans(vi, vn) is true but logpdf is called and not logpdf_with_trans.

I'm not sure, isn't it sufficient to just call `invlink!` after having created the VarInfo object from the template `f.init`? Then the values will be untransformed and the `trans` flag is set to `false`, so we should only get the `logpdf` values.

mohamed82008 commented:
>Then the values will be untransformed and the trans flag is set to false, so we should only get the logpdf values.

But the optimizer is unconstrained, so it may wander off into forbidden territory. We need the decision variables to be unconstrained but the objective to be the `logpdf` of the constrained parameters.

devmotion commented:
That's exactly what I suggested (or at least tried to) - you take the unconstrained values from the optimizer and transform them back to the original space before evaluating the model.

mohamed82008 commented:
> and transform them back to the original space before evaluating the model.

If this is done in an out-of-place manner, then it should be enough yes. But doing it out-of-place is tricky.

devmotion commented:
I guess we could just have an alternative to https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/varinfo.jl#L119-L148 that transforms `x` back instead of just extracting the correct subsets?

devmotion commented:
So basically just changing https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/varinfo.jl#L133.

mohamed82008 commented:
It's tricky because you need to get the transformation correct. If the distribution's support is stochastic, you are out of luck. We don't support stochastic support properly yet but if we do, the context will be the more efficient way to go because it delays the domain transformation until we have the distribution on the RHS of ~ in the `assume` functions.

mohamed82008 commented:
Perhaps a different `AbstractSampler` would also do if it means less methods to define but a context makes more sense since optimization is not sampling.

devmotion commented:
> If the distribution's support is stochastic, you are out of luck.

That's a good point. Currently the transformation is independent of the RHS in the assume statement though: https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/context_implementations.jl#L134

mohamed82008 commented:
> Currently the transformation is independent of the RHS in the assume statement though: https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/context_implementations.jl#L134

Yes which is a bug addressed by https://github.com/TuringLang/DynamicPPL.jl/pull/115. But I will get back to it after NeurIPS.

mohamed82008 commented:
We can do what you say now and fix it later. But I think the context is easier anyways and will likely be faster, no need to `invlink` and `vcat`.

devmotion commented:
Actually, I think we should fix it correctly right away in this PR and define the `OptimizationContext` and the required `assume` and `observe` implementations, which can be very short and optimized I guess.

cpfiffer commented:
I feel like the quicker hack is just to define the following (more or less)

```julia
struct OptimizationSampler <: AbstractSampler end

function assume(
    spl::OptimizationSampler,
    dist::Distribution,
    vn::VarName,
    vi,
)
    if haskey(vi, vn)
        # Always overwrite the parameters with new ones for `SampleFromUniform`.
        if is_flagged(vi, vn, "del")
            unset_flag!(vi, vn, "del")
            r = init(dist, spl)
            vi[vn] = vectorize(dist, r)
            settrans!(vi, false, vn)
            setorder!(vi, vn, get_num_produce(vi))
        else
            r = vi[vn]
        end
    else
        r = init(dist, spl)
        push!(vi, vn, r, dist, spl)
        settrans!(vi, false, vn)
    end
    return r, Bijectors.logpdf_with_trans(dist, r, false)
end
```

cpfiffer commented:
I take that back. I see the benefit of the context, but I don't actually see how you would change this without rewriting `observe` and `assume`. What am I missing as far as implementation is concerned?

trappmartin commented:
> BTW it's new to me but apparently it's a well-known fact that the MAP estimate is not invariant under reparameterization: https://stats.stackexchange.com/questions/330402/what-is-an-example-of-a-transformation-on-a-posterior-distribution-such-that-the

Thanks, I wasn’t aware of that.

devmotion commented:
It would be nice to do something like
```julia
struct OptimizationContext{C}
	context::C
end

# assume
function tilde(ctx::OptimizationContext, sampler, right, vn::VarName, inds, vi)
    # transform variable to original space...
    return tilde(ctx.context, sampler, right, vn, vi)
end

# observe
function tilde(ctx::OptimizationContext, sampler, right, left, vi)
	# transform variable to original space...
    return tilde(ctx.context, sampler, right, left, vi)
end
```
since then one could support arbitrary nested contexts, but that leads to the same AD as transforming everything in the beginning I guess. Alternatively, we could just add special cases for the contexts that we want support by implementing
```
# assume
function tilde(ctx::OptimizationContext, sampler, dist, vn::VarName, inds, vi)
    r = vi[vn]
    return r, Distributions.logpdf(dist, r)
end

# observe
function tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vi)
    return 0
end

function tilde(ctx::OptimizationContext, sampler, dist, value, vi)
    return Distributions.logpdf(dist, value)
end
```
and similar for `dot_tilde`. I'm not sure if we have to consider sampling at all, since optimization requires us to provide the full set of variables anyway I guess? Then one could just use user-provided initial values or sample them once with `SampleFromPrior()` (as currently done), and in the next steps just update these values with the ones provided by the optimization algorithm.

mohamed82008 commented:
I like @devmotion's second suggestion. The `logpdf(NoDist(), ..)` can also be replaced by 0.

devmotion commented:
>  The logpdf(NoDist(..), ..) can also be replaced by 0.

I thought so as well initially, but it seems that it might actually return an array of zeros, depending on the underlying distribution and the provided value. Hence I kept `NoDist`.

mohamed82008 commented:
>I thought so as well initially, but it seems that it might actually return an array of zeros, depending on the underlying distribution and the provided value. Hence I kept NoDist.

I think in all cases where an array is returned, e.g. `dot_assume`, we end up calling `sum` at the end. I think it is safe to assume (pun intended) that the second output of `tilde` and `dot_tilde` will always be a scalar. We call `acclogp!` on it after all.

devmotion commented:
Sounds reasonable, and should also be faster in the cases where `logpdf(::NoDist, x)` would return an array. Do we actually create the array of zeros and sum them up in some place? Sounds a bit inefficient :smile: 

I'll update the code above.

mohamed82008 commented:
>Do we actually create the array of zeros and sum them up in some place? Sounds a bit inefficient 😄

Yes because I didn't want to special case `NoDist` everywhere, it's treated like any other dist. We can probably return a `FillArray`.

devmotion commented:
Maybe an alternative would be to actually work with `loglikelihood` instead of `logpdf` in cases where we always want a scalar, regardless if we have multiple inputs.

cpfiffer commented:
Alright, I'm hitting the mark on @devmotion's math with the `PriorContext`. I'm going to open this up for formal review.

devmotion commented on src/Turing.jl:
```diff
@@ -110,5 +115,8 @@ export  @model,                 # modelling
-        arraydist
+        arraydist,
+
+        MLE,                    # mode estimation tools
+        MAP
```
Can you move these exports to the `@require` block above?

devmotion replied:
You forgot to remove the exports here.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+
+# assume
+function DynamicPPL.tilde(ctx::OptimizationContext{<:LikelihoodContext}, spl, dist, vn::VarName, inds, vi)
+    if haskey(vi, vn)
+        # Always overwrite the parameters with new ones for `SampleFromUniform`.
+        if spl isa SampleFromUniform || is_flagged(vi, vn, "del")
+            unset_flag!(vi, vn, "del")
+            r = init(dist, spl)
+            vi[vn] = vectorize(dist, r)
+            settrans!(vi, false, vn)
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
```
Is
```suggestion
    r = vi[vn]
```
not sufficient?

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, spl, dist, vn::VarName, inds, vi)
+    if haskey(vi, vn)
+        # Always overwrite the parameters with new ones for `SampleFromUniform`.
+        if spl isa SampleFromUniform || is_flagged(vi, vn, "del")
+            unset_flag!(vi, vn, "del")
+            r = init(dist, spl)
+            vi[vn] = vectorize(dist, r)
+            settrans!(vi, false, vn)
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
```
Same here, do we actually need the sampling functionality?
```suggestion
    r = vi[vn]
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    lp = sum(logpdf(dist, r))
```
I guess this should be either
```suggestion
    lp = sum(logpdf.(dist, r))
```
or
```suggestion
    lp = loglikelihood(dist, r)
```
IIRC `logpdf(dist, r)` where `r` are multiple samples is deprecated in Distributions. I'd prefer the latter since it avoid the creation of an intermediate array if not needed (e.g., without AD) but one could still implement custom adjoints based on broadcasting. It seems, however, that we currently always use the first alternative in DynamicPPL, so I guess the custom adjoints are not available right now.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+# dot assume
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:LikelihoodContext}, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
```
Same here, if sampling is not needed, we should really think about implementing `get_val!` that just contains the logic from `get_and_set_val!` for extracting samples.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    lp = sum(logpdf(dist, r))
```
Same here.

trappmartin commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+    init_params = model(f.vi, spl)
+    init_vals = f.vi[spl]
+
+    # Optimize!
```
Could we just use a closure here instead of the additional struct?

torfjelde replied:
I'm also wondering about this. Is there a reason why we don't use something like `make_logjoint` already used for VI: https://github.com/TuringLang/Turing.jl/blob/cb56cd2de7b9f398d5619d78369cd7ee3a2510fb/src/variational/VariationalInference.jl#L230-L257

And for MLE we could just wrap a `LikelihoodCtx`. 

With that being said, the current impl looks nice but might be overly complex with out much gain? :confused:

torfjelde replied:
With current Turing.jl you can do:
```julia
julia> @model function f(x)
           m ~ Normal(0, 1)
           x ~ Normal(m, 1)
       end
DynamicPPL.ModelGen{var"###generator#276",(:x,),(),Tuple{}}(##generator#276, NamedTuple())

julia> model = f(1.5)
DynamicPPL.Model{var"###evaluator#275",(:x,),Tuple{Float64},(),DynamicPPL.ModelGen{var"###generator#276",(:x,),(),Tuple{}}}(##evaluator#275, (x = 1.5,), DynamicPPL.ModelGen{var"###generator#276",(:x,),(),Tuple{}}(##generator#276, NamedTuple()))

julia> b = Turing.bijector(model);

julia> logπ = Turing.Variational.make_logjoint(model);

julia> objective = z -> - (logπ ∘ b)(z);

julia> map = optimize(objective, [0.], LBFGS());

julia> inv(b)(map.minimizer)
1-element Array{Float64,1}:
 0.7499999999941116

julia> map
 * Status: success

 * Candidate solution
    Minimizer: [7.50e-01]
    Minimum:   2.400377e+00

 * Found with
    Algorithm:     L-BFGS
    Initial Point: [1.00e+00]

 * Convergence measures
    |x - x'|               = 2.50e-01 ≰ 0.0e+00
    |x - x'|/|x'|          = 3.33e-01 ≰ 0.0e+00
    |f(x) - f(x')|         = 6.25e-02 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 2.60e-02 ≰ 0.0e+00
    |g(x)|                 = 3.67e-11 ≤ 1.0e-08

 * Work counters
    Seconds run:   0  (vs limit Inf)
    Iterations:    1
    f(x) calls:    3
    ∇f(x) calls:   3
```

You can also get the corresponding ranges by
```julia
julia> b, sym2val = Turing.bijector(model; sym_to_ranges = Val(true));

julia> sym2val
(m = UnitRange{Int64}[1:1],)
```
which could be used internally to construct the `NamedArray`.

torfjelde replied:
Could also use `Turing.gradient_logp` and pass this to `optimize` if there is some magic in there that we need.

devmotion replied:
I suggested doing that and IMO we should change `make_logjoint` etc in the same way. It is much easier to hook into and analyze the internals if we don't use anonymous closures (see e.g. https://github.com/TuringLang/DynamicPPL.jl/issues/45).

torfjelde replied:
That makes sense, and I agree that the closure-idea isn't particularly nice (unless the user does it themselves, making things more transparent) :+1: 

Also, I guess this `OptimizationContext` will cover most use-cases + be slightly more efficient than a closure over `DefaultContext`.

cpfiffer replied:
I'd like to note here that this is more or less what I originally had -- @devmotion requested a callable struct in the first place ([here](https://github.com/TuringLang/Turing.jl/pull/1230#discussion_r413531662)).

I have no preference either way but I don't much care for going back and forth.

cpfiffer replied:
Currently I like @devmotion's original suggestion and would prefer to keep the callable struct. In fact, I think later on it should be generalized outside of the `Optim` context so that everyone can use it for whatever purposes they have.

trappmartin replied:
Make sense.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+    context::C
+    "A `DynamicPPL.VarInfo` struct that will be used to update model parameters."
+    vi::V
+    "Whether the to evaluate the model with transformed variables."
+    linked::Bool
```
I don't think there should be a choice here? `OptimLogDensity` is only used internally, and hence we know that we always optimize in the unconstrained space.

mohamed82008 replied:
I think it's when computing the Hessian that `linked` needs to be false.

cpfiffer replied:
Turns out we don't need to unlink it for the Hessian, since the OptimizationContext handles everything so nicely.

trappmartin replied:
Cool.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+mle = optimize(model, MLE())
+
+# Use a different optimizer
+mle = optimize(model, MLE(), NelderMeade())
```
```suggestion
mle = optimize(model, MLE(), NelderMead())
```
I guess it would be nice to use a julia-repl code block here (see https://docs.julialang.org/en/latest/manual/documentation/, one could also write doctests).

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+map_est = optimize(model, MAP())
+
+# Use a different optimizer
+map_est = optimize(model, MAP(), NelderMeade())
```
Same here:
```suggestion
map_est = optimize(model, MAP(), NelderMead())
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+    return -DynamicPPL.getlogp(varinfo)
+end
+
+"""
+    unlink(f::OptimLogDensity)
+
+Generate an unlinked (with no variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.invlink(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    return OptimLogDensity(f.model, f.context, init, false)
+end
+
+"""
+    link(f::OptimLogDensity)
+
+Generate an linked (with variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.link(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+    return OptimLogDensity(f.model, f.context, init, false)
+end
```
I'm curious, why are `invlink` and `link` needed? BTW the docstring of `invlink` is incorrect.

cpfiffer replied:
I don't think it's needed anymore with `OptimizationContext`, I'll remove it.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+# Various StatsBase methods #
+#############################
+
+function Base.show(io::IO, m::ModeResult)
```
Maybe change this to
```suggestion
function Base.show(io::IO, ::MIME"text/plain", m::ModeResult)
```
and implement another more compact `show` method with just one line (that would be used, e.g., when displaying arrays of ModeResult)? See https://docs.julialang.org/en/v1/manual/types/#man-custom-pretty-printing-1.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+#############################
+
+function Base.show(io::IO, m::ModeResult)
+    println(io, Printf.@sprintf("ModeResult with minimized lp of %.2f", m.lp), "\n")
```
```suggestion
    print(io, "ModeResult with minimized lp of ")
    Printf.@printf(io, "%.2f", m.lp)
    println(io)
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,318 @@
+    varnames, _ = Turing.Inference._params_to_array(ts)
+
+    # Store the parameters and their names in an array.
+    vmat = NamedArrays.NamedArray(vals, varnames)
```
Do you know how it compares to https://github.com/SciML/LabelledArrays.jl?

torfjelde replied:
I haven't tried `NamedArrays` but remember `LabelledArrays` being quite slow when it came to _constructing_ the array, but other than it seemed good. So in this case, `LabelledArrays` should be fine to use :+1:

torfjelde replied:
Potentially interesting though they don't have `LabelledArrays.jl`: https://github.com/mcabbott/AxisKeys.jl/blob/51550fe96f36937f7e4417adf77232f33f35657e/test/speed.jl

cpfiffer replied:
The main reason I picked NamedArrays is because they have good `show` methods. LabelledArrays is pretty damn ugly and not good for presenting results.

torfjelde commented:
Despite being mentioned TWICE, I've completely missed what this was all about. I saw it was a "Draft" and thought "Oh, Cameron is messing about, so I'll have a look with it becomes and actual PR". Did not realize there was a ongoing discussion about the implementation and that I was being pinged several times :flushed:  Sorry guys!

Anyways, first of I think I need to apologize to you @cpfiffer because I think I told you on Slack that for MAP you needed to scale with abs-det-jac but not for MLE, which is just not true (as nicely pointed out by @devmotion earlier!).

So if we want to optimize `p(x, θ)` wrt. `θ`, it does not matter how you arrive at your value `θ` as long it's the one that maximizes `p(x, θ)`. If you want to apply a sequence of transformations `f₁, …, fₙ` to some other variable `ϕ` and then let `θ := (fₙ ∘ ⋯ ∘ f₁)(ϕ)`, that's also fine (as long as you can "reach" the optimal value in `Ω`, it ain't no matter).

So just to clarify the `logabsdetjac` term comes up in either of these cases (mainly):
- in HMC we care about the *expectation* wrt. `p(x, θ) dθ`, i.e. and *integral* over `p(x, θ) dθ`, but after the transformation to ℝᵈ we're instead looking at `p(x, ϕ)` and so we need to rescale the term inside the integral accordingly (i.e. change-of-variables!)
- if we are interested in maximizing `p(x, θ)` where `θ = f(ϕ)` *but* we only have access to `p(x, ϕ)`, then we can derive `p(x, θ)` by scaling with abs-det-jacobian of the transformation `f` (this is the premise of normalizing flows)

Does that make more sense? I'll also have a proper look at the PR now:)

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,267 @@
+
+# dot assume
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:LikelihoodContext}, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
```
`vns` and `dist` are not used, it seems?
```suggestion
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,267 @@
+intended to allow an optimizer to sample in R^n freely.
+"""
+struct OptimizationContext{C<:AbstractContext} <: AbstractContext
+	context::C
```
No tab space:
```suggestion
    context::C
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,267 @@
+# assume
+function DynamicPPL.tilde(ctx::OptimizationContext{<:LikelihoodContext}, spl, dist, vn::VarName, inds, vi)
+    r = vi[vn]
+	return r, 0
```
```suggestion
    return r, 0
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,267 @@
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
```
Why do we use `get_and_set_val!` here but used just `vi[vn]` in the other `dot_tilde` implementations? From looking at the implementation of `get_and_set_val!` (https://github.com/TuringLang/DynamicPPL.jl/blob/f74e039cb7fff30a10be882b6e12e9b686dbc77c/src/context_implementations.jl#L286-L347) I assumed that we need
```julia
function getval(
    vi,
    vns::AbstractVector{<:VarName},
)
    r = vi[vns]
    return r
end

function getval(
    vi,
    vns::AbstractArray{<:VarName},
)
    r = reshape(vi[vec(vns)], size(vns))
    return r
end
```
to just extract the values correctly, and just call `r = getval(vi, vns)` then here. It seems we could even drop the first definition.

cpfiffer replied:
I've added these in for the moment. Seems to be working well.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,267 @@
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = vi[vn]
+    return r, loglikelihood(dist, r)
```
I think we should get rid of the `sum(logpdf.(dist, r))` calls but as I said I assume that just changing them to `loglikelihood` will break AD and be problematic if `dist` is an array of distributions. Maybe we should do that in a separate PR, I guess by introducing a separate `sum_logpdf` in DistributionsAD (to avoid having to commit type piracy when defining `loglikelihood(dist::AbstractArray{<:Distribution}, x)`) with optimized adjoints for AD and a fallback to `loglikelihood(dist::Distribution, x)`.

cpfiffer replied:
Agreed. I'm sure we'll see how this'll fit with AD as people test it out on more complex models.

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,267 @@
+Create a callable `OptimLogDensity` struct that evaluates a model using the given `context`.
+"""
+function OptimLogDensity(model::Model, context::AbstractContext)
+	init = VarInfo(model)
+	DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+	return OptimLogDensity(model, context, init)
```
```suggestion
    init = VarInfo(model)
    DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
    return OptimLogDensity(model, context, init)
```

devmotion commented on test/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,105 @@
+            s ~ InverseGamma(2,3)
+            m ~ Normal(0, sqrt(s))
+        
+            x .~ Normal(m, sqrt(s))
```
`.~` can't be parsed on Julia 1.0

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,267 @@
+function Optim.optimize(model::Model, f::OptimLogDensity, optimizer=Optim.LBFGS(), args...; kwargs...)
+    # Do some initialization.
+    spl = DynamicPPL.SampleFromPrior()
+    init_params = model(f.vi, spl)
```
`init_params` is not used anywhere:
```suggestion
```

devmotion replied:
(and `model(...)` just returns whatever the user returns in the model definition)

trappmartin replied:
Maybe this was supposed to be a `getall(f.vi)`?

cpfiffer commented:
Alright, tests are passing at the moment. Further comments or things you don't think have been addressed yet?

mohamed82008 commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,284 @@
+    f.vi[spl] = M.minimizer
+    invlink!(f.vi, spl)
+    vals = f.vi[spl]
+    link!(f.vi, spl)
```
This means the hessian will not be wrt to the constrained parameters?

cpfiffer replied:
Yup, you're right. This is what I get for relying on bad behavior elsewhere in the code.

mohamed82008 commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,284 @@
+
+function StatsBase.informationmatrix(m::ModeResult; hessian_function=ForwardDiff.hessian, kwargs...)
+    # Calculate Hessian and information matrix.
+    varnames = StatsBase.coefnames(m)
```
Should probably check the vi is invlinked here. If not, then run `invlink!`.

cpfiffer replied:
What's the quick, idiomatic way to check whether the whole VI is linked?

mohamed82008 replied:
islinked(vi, SampleFromPrior())` should be good enough. It will only check the first varname of each symbol though.

cpfiffer replied:
Okay, I made a quick work around to get this to work with `SampleFromPrior`, since `islinked` only works with `Sampler`. I opened a PR to fix this here: https://github.com/TuringLang/DynamicPPL.jl/pull/124

For right now, I think it's fine to use my quick workaround and replace it later if/when the DPPL PR goes through.

cpfiffer commented:
Aight, I think it's finally good to go. :fireworks:

devmotion commented on src/Turing.jl:
```diff
@@ -58,6 +58,11 @@ using .Variational
     end
 end
 
+@init @require Optim="429524aa-4258-5aef-a3af-852621145aeb" @eval begin
+    include("modes/ModeEstimation.jl")
+    export MAP, MLE, optimize
+end
```
Just noticed: users will have problems if they load Optim but not NamedArrays. We should add a nested `@require` block that checks for NamedArrays if we want to keep it optional (which I guess we want). Otherwise one could think about returning a named tuple as default and just provide some optional way for converting it to a NamedArray.

devmotion commented on src/Turing.jl:
```diff
@@ -110,5 +115,8 @@ export  @model,                 # modelling
         NamedDist,
         filldist,
-        arraydist
+        arraydist,
+
+        MLE,                    # mode estimation tools
+        MAP
```
```suggestion
        arraydist
```

devmotion commented on src/modes/ModeEstimation.jl:
```diff
@@ -0,0 +1,301 @@
+import ..DynamicPPL: Model, AbstractContext, VarInfo, AbstractContext, VarName,
+    _getindex, getsym, getfield, settrans!,  setorder!,
+    get_and_set_val!, istrans, tilde, dot_tilde
+import Optim
+import Optim: optimize
+import NamedArrays
```
Optional dependencies have to be loaded as 
```suggestion
import .Optim
import .Optim: optimize
import .NamedArrays
```
otherwise users will get a warning about them not being listed as dependencies of Turing.

mohamed82008 commented:
So if `NamedArrays` is an optional dependency, why does it not have a require block?

cpfiffer commented:
I think I'd actually prefer to make `NamedArrays` a full dependency, just for the moment. I think all this stuff should probably be spun off into a different package so the deps can be managed more specifically.

torfjelde commented:
Also, the VI part would benefit from using `NamedArrays` because right now it's difficult for the user to figure out what the different elements in the vector refers to (you can use `bijector(model; symbol_to_ranges = Val(true))` but that's a "hack" that I've added in so that just have _some_ way of doing it). And the reason why I didn't make a proper solution was to reduce deps. So making `NamedArrays` a explicit dep is also useful for other parts in Turing:)
